### PR TITLE
implement showing password prompt if password not provided

### PR DESCRIPTION
**Pull request by didlich**
implements feature https://github.com/cryptomator/cli/issues/11

### CODE REVIEW

File: src/main/java/org/cryptomator/cli/pwd/PasswordFromStdInputStrategy.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    @Override
+    public String password() {
+        LOG.info("Vault " + "'" + vaultName + "'" + " password from standard input.");
+
+        String password = "";
+        Console console = System.console();
+        if (console == null) {
+            LOG.warn("No console: non-interactive mode, instead use insecure replacement, PW is shown!");
```
Comment (overheadhunter) [inline src/main/java/org/cryptomator/cli/pwd/PasswordFromStdInputStrategy.java]: I like this warning :)

**Comment by overheadhunter** [comment]
Looks good so far. Nice clean style. @markuskreusch will do a full review

File: src/main/java/org/cryptomator/cli/Args.java [hunk 2]
```diff
@@ -133,4 +114,26 @@ public static void printUsage() {
 		new HelpFormatter().printHelp(USAGE, OPTIONS);
 	}
 
+	public PasswordStrategy addPasswortStrategy(final String vaultName) {
```
Comment (markuskreusch) [inline src/main/java/org/cryptomator/cli/Args.java]: Not sure if we actually need this method. I am aware it is used in `CryptomatorCli.validate` but in fact think we should move the whole validation stuff into the Args class. `Args.parse` should fail with exceptions if the provided args are invalid.

Nothing to be done in the related PR though, so OK for now.

File: src/main/java/org/cryptomator/cli/pwd/PasswordFromStdInputStrategy.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+            BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
+            LOG.info("Enter password: ");
```
Comment (markuskreusch) [inline src/main/java/org/cryptomator/cli/pwd/PasswordFromStdInputStrategy.java]: Printing the prompt through a `LOG` call only works well if logging is done to stdout. Maybe we should explicitly use stdout / the console to prompt for the password? Not a problem as long as we do logging the way we do it now. So must not be adressed to merge this.

File: src/main/java/org/cryptomator/cli/pwd/PasswordFromFileStrategy.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+            } catch (IOException e) {
+                return null;
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public void validate() throws IllegalArgumentException {
+        if (!Files.isReadable(pathToFile) || !Files.isRegularFile(pathToFile)) {
```
[the remark this answers is not in the document]
Comment (didlich) [reply src/main/java/org/cryptomator/cli/pwd/PasswordFromFileStrategy.java]: OK, if there is something you prefer, so tell me.
This condition is currently used in:
https://github.com/cryptomator/cli/blob/7bf9dc06af5985a9afead5a9689ceb03f2088744/src/main/java/org/cryptomator/cli/Args.java#L115
so this line is just copied and negated

File: src/main/java/org/cryptomator/cli/pwd/PasswordFromStdInputStrategy.java (hunk 3 above)
Comment (didlich) [reply src/main/java/org/cryptomator/cli/pwd/PasswordFromStdInputStrategy.java]: You are right, I will change this.

**Comment by didlich** [comment]
@markuskreusch I want to ask you about the time point the password is entered in the case password is read from standard input
in the current implementation the user is asked to input the password shortly before vault is decrypted
I wonder if we should ask the password in the args processing step, it has the drawback that the password is stored in memory the whole time the cli is running, but has the advantage all the passwords are available before we start processing the vaults.

What do you think?

File: src/main/java/org/cryptomator/cli/pwd/PasswordFromFileStrategy.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (overheadhunter) [reply src/main/java/org/cryptomator/cli/pwd/PasswordFromFileStrategy.java]: > Remeber what we currently use as replacement for isRegularFile?

We're using `Files.isReadable()` now.

**Comment by overheadhunter** [comment]
> it has the drawback that the password is stored in memory the whole time the cli is running

Of course we try to have the password as short a time as possible in memory, but even afterwards the derived key would be in memory for as long as the vault is unlocked. An attacker able to create memory dumps would still be able to use this key to decrypt the data. In other words, if the machine is already compromised, all is lost.

So in my opinion it is acceptable to have the password in memory for a longer period, if it serves a purpose.

**Comment by didlich** [comment]
@markuskreusch, @overheadhunter hi, could you have a look at the changes

**Comment by markuskreusch** [comment]
I am really sorry for the delay. Forgot about this during the holidays.
Looks fine now.

**Pull request merged**